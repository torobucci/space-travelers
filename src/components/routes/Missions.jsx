import { useSelector, useDispatch } from 'react-redux';
import { React } from 'react';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import { joinMission, leaveMission } from '../redux/missions/missionsSlice';

const Missions = () => {
  const { missions } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  return (
    <Table striped bordered className="container">
      <thead>
        <tr>
          <th>Missions</th>
          <th>Description</th>
          <th>Status</th>
          <th aria-label="none" />
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={mission.mission_id}>
            <td>{mission.mission_name}</td>
            <td>{mission.description}</td>
            {mission.reserved && (
            <>
              <td className="align-middle"><Button variant="primary" size="sm" className="not-member">ACTIVE MEMBER</Button></td>
              <td className="align-middle"><Button variant="outline-danger" size="sm" className="join-mission" onClick={() => dispatch(leaveMission(mission.mission_id))}>Leave Mission</Button></td>
            </>
            )}
            {!mission.reserved && (
            <>
              <td className="align-middle"><Button variant="secondary" size="sm" className="not-member">NOT A MEMBER</Button></td>
              <td className="align-middle"><Button variant="outline-secondary" size="sm" className="join-mission" onClick={() => dispatch(joinMission(mission.mission_id))}>Join Mission</Button></td>
            </>
            )}
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

export default Missions;
