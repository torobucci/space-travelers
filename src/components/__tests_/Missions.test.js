import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import Missions from '../routes/Missions';
import '@testing-library/jest-dom';

jest.mock('axios', () => 'Mockedaxios');
const mockStore = configureStore([]);

describe('Missions component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      missions: {
        missions: [
          {
            mission_id: 1,
            mission_name: 'Mission 1',
            description: 'Mission 1 Description',
            reserved: false,
          },
          {
            mission_id: 2,
            mission_name: 'Mission 2',
            description: 'Mission 2 Description',
            reserved: true,
          },
        ],
      },
    });

    component = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
  });

  it('should render the table with the correct headers', () => {
    const { getByText } = component;
    expect(getByText('Missions')).toBeInTheDocument();
    expect(getByText('Description')).toBeInTheDocument();
    expect(getByText('Status')).toBeInTheDocument();
  });

  it('should render the table with the correct data', () => {
    const { getByText } = component;
    expect(getByText('Mission 1')).toBeInTheDocument();
    expect(getByText('Mission 1 Description')).toBeInTheDocument();
    expect(getByText('NOT A MEMBER')).toBeInTheDocument();
    expect(getByText('Join Mission')).toBeInTheDocument();
    expect(getByText('Mission 2')).toBeInTheDocument();
    expect(getByText('Mission 2 Description')).toBeInTheDocument();
    expect(getByText('ACTIVE MEMBER')).toBeInTheDocument();
    expect(getByText('Leave Mission')).toBeInTheDocument();
  });

  it('should dispatch joinMission action when Join Mission button is clicked', () => {
    const { getByText } = component;
    fireEvent.click(getByText('Join Mission'));
    const actions = store.getActions();
    expect(actions[0].type).toBe('missions/joinMission');
    expect(actions[0].payload).toBe(1);
  });

  it('should dispatch leaveMission action when Leave Mission button is clicked', () => {
    const { getByText } = component;
    fireEvent.click(getByText('Leave Mission'));
    const actions = store.getActions();
    expect(actions[0].type).toBe('missions/leaveMission');
    expect(actions[0].payload).toBe(2);
  });
  it('Missions snapshots match', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
