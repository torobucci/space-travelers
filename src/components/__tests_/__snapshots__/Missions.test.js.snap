// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component Missions snapshots match 1`] = `
<table
  className="container table table-striped table-bordered"
>
  <thead>
    <tr>
      <th>
        Missions
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th
        aria-label="none"
      />
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Mission 1
      </td>
      <td>
        Mission 1 Description
      </td>
      <td
        className="align-middle"
      >
        <button
          className="not-member btn btn-secondary btn-sm"
          disabled={false}
          type="button"
        >
          NOT A MEMBER
        </button>
      </td>
      <td
        className="align-middle"
      >
        <button
          className="join-mission btn btn-outline-secondary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Mission 2
      </td>
      <td>
        Mission 2 Description
      </td>
      <td
        className="align-middle"
      >
        <button
          className="not-member btn btn-primary btn-sm"
          disabled={false}
          type="button"
        >
          ACTIVE MEMBER
        </button>
      </td>
      <td
        className="align-middle"
      >
        <button
          className="join-mission btn btn-outline-danger btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Leave Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
